### DISCUSSION TITLE
Visual display refreshes slower in lazer than in osu! stable

### DISCUSSION THREAD

**Discussion by abraker95** [ppy/osu#16375] [Q&A]
I am not using the word "frame rate" in the title because lazer's reported frame rate doesn't seem to matching what is actually being displayed. My display runs on 165 Hz and have set lazer on 2x refresh rate. Since lazer reports a draw rate of 330 Hz I shouldn't see a difference between lazer or osu!, yet this is not the case.

I have taken slow-mo footage using my Samsung Galaxy S10 (at 480 fps) to compare how notes in mania scroll down in lazer and stable.
[osu lazer gameplay slow mo](https://www.youtube.com/watch?v=roDh93iPNOw)
[osu stable gameplay slow mo](https://www.youtube.com/watch?v=dvNB9OKPGuQ)

And have also taken separate videos of frame graphs. I didn't go with screen capture as to avoid any frame rate issues due to recording.
[osu lazer frame graph](https://www.youtube.com/watch?v=VSnTIQL0SPU)
[osu stable frame graph](https://www.youtube.com/watch?v=URnvwkRqFgk)

__PC Specs__
OS: Win 10
Graphics card: AMD Radeon HD 7700 Series
Monitor: Sceptre E255B-1658A

**Comment by integerrr** [comment]
Well first off lazer didn't report 330 draw fps when it scrolls
![image](https://user-images.githubusercontent.com/7279624/148668173-25287621-95a0-4094-8f7f-57cdb80e1a15.jpeg)

**Reply by abraker95** [reply]
You are looking at a point in time when it is on the failed screen/menu, not during gameplay. Look at anything before the 6 second mark.

**Comment by bdach** [comment]
i don't know how to interpret the above videos. as far as i can tell this could as well be caused by a scroll speed mismatch.

the following would be helpful, if you are able to perform it:

* start mania in autoplay
* press F1 to open the draw visualiser
* hover "choose target"
* select any gameplay element and click left mouse
* scroll the visualiser window down to show the "Time" property
* capture a video of how that changes and attach it here

the goal is to see what the interval between displayed frames is, independent of things like mismatching scroll speed.

**Reply by abraker95** [reply]
For the second step I had to press ctrl + F1, not just F1, to open the draw visualizer.

Here is the slow-mo footage with the time property shown of one of the mania columns. Also the draw visualizer decreased the draw FPS to around 200 while it was being displayed: [video](https://www.youtube.com/watch?v=_sBGCmjH1kk)

**Reply by bdach** [reply]
well that video is showing 15-20ms frame times. which looks a lot like 60fps/vsync.

also, for sanity's sake, is this exclusive fullscreen or borderless?

**Reply by abraker95** [reply]
Borderless

**Reply by bdach** [reply]
does exclusive fullscreen change anything?

**Reply by abraker95** [reply]
It does, displayed rate is what I expect it to be

**Reply by bdach** [reply]
interesting. good thing that at least exclusive fullscreen works as expected, but narrowing down why borderless doesn't is going to be more painful. there are a few suspects, including SDL, windows itself or something device/driver-specific.

**Reply by abraker95** [reply]
I've tried windowed mode and it seems to work fine. This issue seems to be exclusive to borderless.

**Reply by peppy** [reply]
Yes, because it's vsync at fault. Do you not believe me? I'm not sure I understand what you want here.

Windowed mode will be running using the DWM's buffering, which will guarantee correct vsync. When you run in a fullscreen mode, it's up to the GPU driver to handle things correctly, which it's not doing for whatever reason.

Try different driver versions, reset any custom settings you have, and google for solutions for your hardware/driver combination.

**Comment by peppy** [comment]
Your'e running stable on unlimited? Can you compare apples to apples?

Looks liek a bog-standard vsync mismatch to me. You can see lazer sometimes hitting the vsync interval when and sometimes missing it.

Are you not using gsync/freesync?

**Reply by abraker95** [reply]
> Your'e running stable on unlimited? Can you compare apples to apples?

My understanding is while stable is running on unlimited I should see the display refresh as fast as my screen allows, which is 165 Hz. Lazer reports 330 Hz so I should see the display refresh at 165 Hz as well. So theoretically there should not be significantly observable difference between the two.

I can take slow-mo footage of stable running on a different frame limiter setting if you believe that will help figure out the problem (which setting if so?).

> Are you not using gsync/freesync?

I am not. My GPU software says it's not supported

![](https://i.imgur.com/59uTNi8.png)

**Reply by peppy** [reply]
please test what i said. set lazer to either vaunt or unlimited.

**Reply by abraker95** [reply]
I set lazer to unlimited and display did not refresh any faster

**Reply by peppy** [reply]
and vsync?

**Reply by abraker95** [reply]
Same as unlimited; does not refresh any faster

**Reply by peppy** [reply]
No idea then. Something in your GPU pipeline is not swapping buffers in time. Maybe try different driver versions, and make sure you don't have any third party overlays that could be holding things up.

**Reply by abraker95** [reply]
That doesn't make sense to me. Why would stable work fine but lazer not?

**Reply by peppy** [reply]
Different graphic stacks. Different gl version. Lots different. Anything could be the issue.

All I can tell you is the cause: your driver is not swapping buffers when instructed to. There's a sync issue somewhere on your system. A solution to this is out of scope of what we can offer here.

**Comment by Susko3** [comment]
Try disabling full screen optimizations:
https://devblogs.microsoft.com/directx/demystifying-full-screen-optimizations/ (Scroll down to `Troubleshooting`)

**Reply by abraker95** [reply]
This issue is specific to borderless, not full screen. Well regardless, I tried your suggestion and it did not fix the issue.
